design/ErrorHandling.md

```diff
@@ -7,75 +7,117 @@ There are a number of cases that developers or applications need error handling
 
 ## *Debugging*: Dev Tools
 
 Implementations should provide a way to enable synchronous validation, for example via a debug shim or via the developer tools.
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
-## *Telemetry* \& *Fallback*: Error Logging
+The behavior is implementation-defined. For example, it may throw an exception, or it may just cause the debugger to pause inside the call that fails.
 
-`WebGPUDevice` keeps a log of errors that happened on each thread (main thread or Web Worker).
-The application "eventually" recieves these errors, asynchronously via a callback.
+## *Fatal Errors*: Lost/Recovered Events
 
-These errors are all delivered through one stream - not directly associated with the objects or operations that produced them.
-They should not be used by applications to recover from expected, recoverable errors.
-Instead, the error log may be used for handling unexpected errors in deployment, for bug reporting and telemetry.
+TODO: distinguish device and adapter loss
+TODO: consider case: unplug discrete and switch to integrated
+TODO: consider case: multiple "apps" on one page using different devices
 
-```
-enum WebGPULogEntryType {
-    "device-lost",
-    "validation-error",
-    "recoverable-out-of-memory",
+```webidl
+enum GPUEventType {
+    "gpudevicelost",
+    "gpudevicerecovered"
 };
 
-interface WebGPULogEntry {
-    readonly attribute WebGPULogEntryType type;
-    readonly attribute any object;
+interface GPUDeviceLostEvent : Event {
+    readonly attribute bool unrecoverable;
     readonly attribute DOMString? reason;
 };
 
-callback WebGPULogCallback = void (WebGPULogEntry error);
+interface GPUDeviceRecoveredEvent : Event {};
+
+partial interface GPUDevice : EventTarget {};
+```
+
+```js
+device.addEventListener("gpudevicelost", (/*GPUDeviceLostEvent*/ event) => {
+    event.preventDefault();
+});
+
+device.addEventListener("gpudevicerecovered", (/*GPUDeviceRecoveredEvent*/ event) => {
+    // Create a new GPUDevice. (The old one is not valid.)
+});
+```
+
+ - `"gpudevicelost"`
+    Signals that the GPUDevice cannot be used anymore.
+    Any further operations on the device, or objects associated with the device, are errors.
+    This may happen if the device is reclaimed by the browser,
+    or something goes fatally wrong (e.g. crash or native device loss).
+    (TODO: does it happen if the device is destroyed by the application?)
+
+ - `"gpudevicerecovered"`
+    Signals when it is okay for the application to make a new GPUDevice.
+    The original device (the event target) is still invalid - it does not become valid.
+    `"gpudevicerecovered"` may only fire if:
+     - the device has received a `"gpudevicelost"` event,
+     - the `"gpudevicelost"` event's `"unrecoverable"` flag was unset, and
+     - the `"gpudevicelost"` event's default action was not taken (i.e. `preventDefault` was called).
```

Comment:
Do you know why an application would choose to not call `preventDefault`? I'm also not sure what the browser can do differently when `preventDefault` is not called.

Reply:
I spoke with Ken and it seems that preventDefault is only really necessary for webglcontextlost because it restores the context on the same WebGLRenderingContext object. It signals to the browser that the app knows it's handling this case, and not just blindly continuing to issue GL calls to the same context.

I'm removing it.